homeassistant/components/media_player/kodi.py

```diff
@@ -37,82 +72,218 @@
+                if self._websocket:
+                    yield from self._websocket.close()
+                yield from self._session.close()
+        except (aiohttp.errors.ClientError,
+                asyncio.TimeoutError,
+                ConnectionRefusedError):
+            pass
 
     @property
     def name(self):
         """Return the name of the device."""
         return self._name
 
-    def _get_players(self):
-        """Return the active player objects or None."""
-        import jsonrpc_requests
+    @asyncio.coroutine
+    def async_json_request(self, method, *args):
+        """Request json information from Kodi.
+
+        Returns None if connection failed.
+        """
+        data = {
+            ATTR_JSONRPC: JSONRPC_VERSION,
+            ATTR_METHOD: method,
+            ATTR_ID: str(uuid.uuid4()),
+        }
+        if len(args) > 0:
+            data[ATTR_PARAMS] = args
+
+        response = None
         try:
-            return self._server.Player.GetActivePlayers()
-        except jsonrpc_requests.jsonrpc.TransportError:
-            if self._players is not None:
-                _LOGGER.warning('Unable to fetch kodi data')
-                _LOGGER.debug('Unable to fetch kodi data', exc_info=True)
+            with async_timeout.timeout(DEFAULT_TIMEOUT, loop=self.hass.loop):
+                response = yield from self._session.post(
+                    self._http_url, data=json.dumps(data))
+
+            if response.status == 401:
+                _LOGGER.error(
+                    "Error fetching Kodi data. HTTP %d Unauthorized. "
+                    "Password is incorrect.", response.status)
+                return None
+            if response.status != 200:
+                _LOGGER.error(
+                    "Error fetching Kodi data. HTTP %d", response.status)
+                return None
+
+            response_json = yield from response.json()
+            if 'error' in response_json:
+                _LOGGER.error(
+                    "RPC Error Code %d: %s",
+                    response_json['error']['code'],
+                    response_json['error']['message'])
+                return None
+            return response_json['result']
+        except (aiohttp.errors.ClientError,
+                asyncio.TimeoutError,
+                ConnectionRefusedError):
             return None
+        finally:
+            if response:
+                response.close()
+
+    @asyncio.coroutine
+    def async_websocket_loop(self):
+        """Check websocket for push events from Kodi."""
+        import websockets
+
+        while True:
+            try:
+                msg = yield from self._websocket.recv()
+            except websockets.exceptions.ConnectionClosed:
+                break
+
+            data = json.loads(msg)
+            if data[ATTR_METHOD] in EXIT_NOTIFICATIONS:
+                self._players = None
+                yield from self.async_update_ha_state()
```

Comment:
You don't have to wait for the state to update before processing the next message. Consider using `hass.async_add_job(self.async_update_ha_state())`